src/modules/cache.rs

```diff
@@ -0,0 +1,123 @@
+/// multiple noise modules. If a source module is not cached, the source
+/// module will redundantly calculate the same output value once for each
+/// noise module in which it is included.
+#[derive(Clone, Debug)]
+pub struct Cache<Source, T> {
+    /// Outputs the value to be cached.
+    pub source: Source,
+
+    value: RefCell<Option<T>>,
+
+    point: RefCell<Vec<T>>,
+}
+
+impl<Source, T> Cache<Source, T> {
+    pub fn new(source: Source) -> Cache<Source, T> {
+        Cache {
+            source: source,
+            value: RefCell::new(None),
+            point: RefCell::new(Vec::new()),
+        }
+    }
+}
+
+impl<Source, T> NoiseModule<Point2<T>> for Cache<Source, T>
+    where Source: NoiseModule<Point2<T>, Output = T>,
+          T: Float,
+{
+    type Output = T;
+
+    fn get(&self, point: Point2<T>) -> Self::Output {
+        let cached_point: Vec<T>;
+        let cached_value: Option<T>;
+        {
+            cached_point = self.point.borrow().clone();
+            cached_value = self.value.borrow().clone();
+        }
+
+        if cached_value == None || point[0] != cached_point[0] || point[1] != cached_point[1] {
+            *self.value.borrow_mut() = Some(self.source.get(point));
+            *self.point.borrow_mut() = point.iter().cloned().collect::<Vec<_>>();
+        }
+
+        return self.value.borrow().unwrap();
+    }
+}
+
+impl<Source, T> NoiseModule<Point3<T>> for Cache<Source, T>
+    where Source: NoiseModule<Point3<T>, Output = T>,
+          T: Float,
+{
+    type Output = T;
+
+    fn get(&self, point: Point3<T>) -> Self::Output {
+        let cached_point: Vec<T>;
+        let cached_value: Option<T>;
+        {
+            cached_point = self.point.borrow().clone();
+            cached_value = self.value.borrow().clone();
+        }
+
+        if cached_value == None || point[0] != cached_point[0] || point[1] != cached_point[1] ||
+           point[2] != cached_point[2] {
+            *self.value.borrow_mut() = Some(self.source.get(point));
+            *self.point.borrow_mut() = point.iter().cloned().collect::<Vec<_>>();
+        }
+
+        return self.value.borrow().unwrap();
+    }
+}
+
+impl<Source, T> NoiseModule<Point4<T>> for Cache<Source, T>
+    where Source: NoiseModule<Point4<T>, Output = T>,
+          T: Float,
+{
+    type Output = T;
+
+    fn get(&self, point: Point4<T>) -> Self::Output {
+        let cached_point: Vec<T>;
+        let cached_value: Option<T>;
+        {
+            cached_point = self.point.borrow().clone();
+            cached_value = self.value.borrow().clone();
+        }
+
+        if cached_value == None || point[0] != cached_point[0] || point[1] != cached_point[1] ||
+           point[2] != cached_point[2] ||
+           point[3] != cached_point[3] {
+            *self.value.borrow_mut() = Some(self.source.get(point));
+            *self.point.borrow_mut() = point.iter().cloned().collect::<Vec<_>>();
```

Comment:
I already tried that. point is not a Vec, it's an array on the stack. It doesn't like being converted into a Vec, so I had to go through these shenanigans to convert it from an array to a Vec.